### PR TITLE
Support reading and writing intra-residue bonds in PDBx files 

### PR DESCRIPTION
**Pull request by padix-key**
This PR add support for reading/writing intra-residue bonds via the ``chem_comp_bond`` category in PDBx files. Files that contain this category were introduced in the [PDB NextGen Archive](https://www.wwpdb.org/ftp/pdb-nextgen-archive-site).

### CODE REVIEW

**Review by cisert — CHANGES_REQUESTED** [review]
I think a pretty common use case might be reading SDF files generates from some cheminformatics software (e.g., RDKit), and then wanting to have the same bond information in cif format. 
I've tried reading an RDKit-generated SDF file (attached below, rename to `aspirin.sdf`), saving it as PDBx (incl. bonds), then reading it again from cif, which didn't seem to keep the bond information: 

```python 
import biotite.structure.io as strucio
import biotite.structure.io.pdbx as pdbx

arr = strucio.load_structure("aspirin.sdf")
print(arr.bonds.as_array().shape) # (13, 3)

f = pdbx.CIFFile()
pdbx.set_structure(f, arr, include_bonds=True)
f.write("/Users/cisert/Downloads/aspirin.cif")

arr2 = strucio.load_structure("aspirin.cif", include_bonds=True)
print(arr2.bonds.as_array().shape) # (0, 3)
```
Am I missing something? 

[aspirin.txt](https://github.com/biotite-dev/biotite/files/15466289/aspirin.txt)

File: src/biotite/structure/bonds.pyx [hunk 1]
```diff
@@ -1562,7 +1563,12 @@ def connect_via_residue_names(atoms, atom_mask=None, bint inter_residue=True):
     inter_residue : bool, optional
         If true, connections between consecutive amino acids and
         nucleotides are also added.
-
+    custom_bond_dict : dict (str -> dict ((str, str) -> int)), optional
+        A dictionary of dictionaries:
+        The inner dictionary maps tuples of two atom names
+        to their respective bond types (represented as integer) for .
```
[the remark this answers is not in the document]
Comment (padix-key) [reply src/biotite/structure/bonds.pyx:1569]: True, thanks for spotting this

**Comment by padix-key** [comment]
I reproduced your problem. There are two issues here:

- The intra-bond parsing code in `get_structure()` was placed so it only ran for `AtomArrayStack`. The latest commit fixed this.
- SDF files do neither contain residue nor atom names, so `res_name` and `atom_name` are empty in the parsed `AtomArray`. However `chem_comp_bond` in PDBx requires these two columns for identifying bonded atoms. Hence, you need to assign them before writing the CIF file:
  ```python
  import biotite.structure.io as strucio
  import biotite.structure.io.pdbx as pdbx

  arr = strucio.load_structure("aspirin.sdf")
  arr.res_name[:] = "ASP"
  # Probably you would want a better atom naming scheme
  arr.atom_name[:] = [f"A{i}" for i in range(arr.array_length())]

  f = pdbx.CIFFile()
  pdbx.set_structure(f, arr, include_bonds=True)
  f.write("aspirin_temp.cif")

  arr2 = strucio.load_structure("aspirin_temp.cif", include_bonds=True)
  ```

**Comment by cisert** [comment]
Did the code snippet you posted above run for you? Pulling your changes and running locally, I got an error from the final line 
`TypeError: connect_via_residue_names() got an unexpected keyword argument 'custom_bond_dict'`

**Comment by padix-key** [comment]
> Did the code snippet you posted above run for you? Pulling your changes and running locally, I got an error from the final line `TypeError: connect_via_residue_names() got an unexpected keyword argument 'custom_bond_dict'`

Did you reinstall the package? `connect_via_residue_names()` is part of a Cython module. Thus, the compiled module will still have the old version, if not explicitly reinstalled.

**Pull request merged**